packages/ui/src/components/Form/customField/TypeaheadField.tsx

```diff
@@ -0,0 +1,239 @@
+  let allOptions: SelectOptionProps[] = useMemo(() => {
+    const hasSelectedReference = props.options?.some((option) => option.value === selectedReference);
+    if (!hasSelectedReference && selectedReference !== '') {
+      const newOption: SelectOptionObject = {
+        value: selectedReference,
+      };
+      props.options?.push(newOption);
+    }
+    return props.options!.map((option) => {
+      return {
+        value: option.value,
+        children: option.value,
+        isSelected: option.value === selectedReference,
+      };
+    });
+  }, [selectedReference]);
+
+  const [isOpen, setIsOpen] = useState(false);
+  const [selected, setSelected] = useState<string>('');
+  const [inputValue, setInputValue] = useState<string>(selectedReference);
+  const [filterValue, setFilterValue] = useState<string>('');
+  const [selectOptions, setSelectOptions] = useState<SelectOptionProps[]>(allOptions);
+  const [focusedItemIndex, setFocusedItemIndex] = useState<number | null>(null);
+  const [activeItem, setActiveItem] = useState<string | null>(null);
+  const [onCreation, setOnCreation] = useState<boolean>(false); // Boolean to refresh filter state after new option is created
+  const textInputRef = useRef<HTMLInputElement>();
+
+  useEffect(() => {
+    let newSelectOptions: SelectOptionProps[] = [...allOptions];
+
+    // Filter menu items based on the text input value when one exists
+    if (filterValue) {
+      newSelectOptions = allOptions.filter((menuItem) =>
+        String(menuItem.children).toLowerCase().includes(filterValue.toLowerCase()),
+      );
+      // When no options are found after filtering, display creation option
+      if (!newSelectOptions.length) {
+        newSelectOptions = [{ isDisabled: false, children: `Insert custom value "${filterValue}"`, value: 'create' }];
+      }
+      // Open the menu when the input value changes and the new value is not empty
+      if (!isOpen) {
+        setIsOpen(true);
+      }
+    }
+    setSelectOptions(newSelectOptions);
+    setActiveItem(null);
+    setFocusedItemIndex(null);
+  }, [filterValue, onCreation, isOpen]);
+
+  const onToggleClick = () => {
+    setIsOpen(!isOpen);
+  };
+
+  const onSelect = (_event: React.MouseEvent<Element, MouseEvent> | undefined, value: string | number | undefined) => {
+    if (value) {
+      if (value === 'create') {
+        if (!allOptions.some((item) => item.value === filterValue)) {
+          allOptions = [...allOptions, { value: filterValue, children: filterValue }];
+        }
+        setSelected(filterValue);
+        setOnCreation(!onCreation);
+        props.onChange(filterValue);
+        setFilterValue('');
+      } else {
+        setInputValue(value as string);
+        setFilterValue('');
+        setSelected(value as string);
+        props.onChange(value as string);
+      }
+    }
+    setIsOpen(false);
+    setFocusedItemIndex(null);
+    setActiveItem(null);
+  };
```

Comment:
[curious-question]: Wouldn't this nested `if` be simplified as:

```typescript
    if (value === 'create') {
        if (!allOptions.some((item) => item.value === filterValue)) {
          allOptions = [...allOptions, { value: filterValue, children: filterValue }];
        }
        setSelected(filterValue);
        setOnCreation(!onCreation);
        props.onChange(filterValue);
        setFilterValue('');
      } else {
        setInputValue(value as string);
        setFilterValue('');
        setSelected(value as string);
        props.onChange(value as string);
      }
    }
    setIsOpen(false);
    setFocusedItemIndex(null);
    setActiveItem(null);
  };
```

Reply:
yes, you're right, the first check was redundant. Updated.